README.md
My greasemonkey scripts
=======================

- http://spenibus.net
- https://github.com/spenibus/greasemonkey-scripts
- https://gitlab.com/spenibus/greasemonkey-scripts


escapistmagazine
----------------

- video download links.


hitbox
------

- archive download links
    - link text is formatted (a download manager such as DownThemAll can use that info for the filename)
    - most of the link text is hidden for readability but remains useable regarding the previous point


jeuxvideocom
------------

- video download links


stackoverflow
-------------

- flags stats


twitch
------

- quick links to popout and chat
- live playlist links
- archive download links
    - link text is formatted (a download manager such as DownThemAll can use that info for the filename)
    - most of the link text is hidden for readability but remains useable regarding the previous point


youtube
-------

- disables ytspf (ajax page loading & history.pushState)
- video download links
    - link text is formatted (a download manager such as DownThemAll can use that info for the filename)
    - some videos are "protected" and download will fail (I'm not annoyed enough to look into it)
- proxfree links
- sets wmode of embedded flash player to "direct"
    - accelerated video rendering
    - no hiccup on fullscreen switching

changelog.txt
2015-09-23 02:55 +0000
- stackoverflow: separate processed flags stats


2015-08-07 04:13 +0000
- added stackoverflow
- stackoverflow: flags stats


2015-05-04 02:05 +0000
- added escapistmagazine


2015-05-04 02:05 +0000
- updated lib


2015-03-03 13:59 +0000
- twitch: updated vod processing


2015-02-14 17:28 +0000
- jeuxvideocom: updated id parsing


2015-01-25 19:45 +0000
- twitch: added vod download size estimation


2015-01-22 05:29 +0000
- twitch: css update


2015-01-21 04:34 +0000
- twitch: version bump


2015-01-21 04:32 +0000
- twitch: vod links titles additional typo fix


2015-01-21 04:14 +0000
- twitch: vod links titles typo fix


2015-01-21 04:10 +0000
- twitch: updated vod links titles


2015-01-21 04:00 +0000
- twitch: updated new vod download


2015-01-21 01:52 +0000
- twitch: updated new vod master playlist parsing


2014-12-18 20:36 +0000
- twitch: updated playlist fetching


2014-12-14 19:45 +0000
- added jeuxvideocom


2014-12-08 21:00 +0000
- twitch: added rudimentary /v/ url support
- twitch: added fallback for /v/ url when /b/ is available as referrer
- added changelog


2014-11-22 02:29 +0000
- updated readme again


2014-11-22 02:28 +0000
- updated readme


2014-11-22 02:27 +0000
- added hitbox, for archive download


2014-11-18 12:51 +0000
- twitch/youtube: updated event listeners to DOMContentLoaded


2014-10-31 22:37:35 +0000
- fixed minor var misuse


2014-10-31 22:31 +0000
- added muted chunks count
- css update
- rewrote channel name acquisition


2014-10-30 04:10 +0000
- youtube: modified link text format


2014-10-28 18:27 +0000
- twitch: updated version number, it was forgotten in previous commit


2014-10-28 18:24 +0000
- twitch: added notification for muted video chunks


2014-10-25 00:47 +0000
- youtube: show wmode notification on watch page only


2014-10-24 23:20 +0000
- added youtube script
- updated version numbering to use UTC
- updated readme


2014-10-18 03:32 +0000
- readme update


2014-10-18 03:27 +0000
- initial commit: lib + twitch

hitbox.user.js
// ==UserScript==
// @name        hitbox
// @namespace   greasemonkey@spenibus
// @include     http*://hitbox.tv/*
// @include     http*://*.hitbox.tv/*
// @version     20141122-0224
// @require     spenibus-greasemonkey-lib.js
// @grant       unsafeWindow
// @grant       GM_xmlhttpRequest
// ==/UserScript==


/*******************************************************************************
creation: 2010-11-21 23:11 +0000
  update: 2014-11-22 02:24 +0000
*******************************************************************************/




/***************************************************************** shorthands */
var SGL = spenibus_greasemonkey_lib;
var loc = document.location;




/**************************************************************** format time */
function timeFormat(timestamp) {
   if(timestamp) {
      var date = new Date(timestamp);
      return date.getFullYear()+'-'+
         ('0'+(date.getMonth()+1)).slice(-2)+'-'+
         ('0'+date.getDate()).slice(-2)+' '+
         ('0'+date.getHours()).slice(-2)+'-'+
         ('0'+date.getMinutes()).slice(-2)+'-'+
         ('0'+date.getSeconds()).slice(-2);
   }
   return 0;
}




/************************************** format duration from seconds to h:m:s */
function durationFormat(secs) {
   var h = Math.floor(secs/3600);
   var m = ('0'+Math.floor(secs%3600/60)).slice(-2);
   var s = ('0'+(secs%3600%60)).slice(-2);
   return h+':'+m+':'+s;
}




/*******************************************************************************
********************************************************************************
********************************************************************************
************************************************************** reload buttons */
function reloadButtons() {


   SGL.css(''
      // hide by default
      +'#spenibusReload {'
         +'display:none;'
      +'}'
      // show on container hover
      +':hover > #spenibusReload {'
         +'display:block;'
      +'}'
      // button
      +'#spenibusReload button {'
         +'font-weight:bold;'
         +'margin:4px;'
         +'padding:0 4px;'
      +'}'
   );

   var box = SGL.displayBox('spenibusReload');

   // button name : function
   var list = {
      'reload archives' : archives,
   }

   box.set('');
   for(var i in list) {
      var button = document.createElement('button');
      button.textContent = i;
      button.addEventListener('click', list[i], false);
      box.node.appendChild(button);
   }
}
window.addEventListener('DOMContentLoaded', reloadButtons, false);




/*******************************************************************************
********************************************************************************
********************************************************************************
********************************************************************* archive */
function archives() {


   SGL.css(''
      +'#spenibus_archive {'
         +'max-height:400px;'
         +'overflow-y:scroll;'
         +'margin:0;'
         +'padding:0;'
      +'}'
      +'#spenibus_archive * {'
         +'margin:0;'
         +'padding:0;'
      +'}'
      +'#spenibus_archive > .list {'
         +'display:table;'
         +'white-space:nowrap;'
      +'}'
      +'#spenibus_archive > .list > .row {'
         +'display:table-row;'
      +'}'
      +'#spenibus_archive > .list > .row > * {'
         +'display:table-cell;'
         +'padding:0px 2px;'
         +'text-align:right;'
      +'}'
      // cells: grid
      +'#spenibus_archive > .list > .row + .row > * {'
         +'border-top:1px solid #888;'
      +'}'
      +'#spenibus_archive > .list > .row > * + * {'
         +'border-left:1px solid #888;'
      +'}'
      // show only first cell of header row by default
      +'#spenibus_archive > .list > .row > * {'
         +'display:none;'
      +'}'

      +'#spenibus_archive > .list > .row:nth-of-type(1) > *:nth-of-type(1) {'
         +'display:table-cell;'
      +'}'
      // show everything on container hover
      +':hover > #spenibus_archive > .list > .row > * {'
         +'display:table-cell;'
      +'}'
      // hide extra link text
      +'#spenibus_archive .extra {'
         +'display:none;'
      +'}'
      // row hover
      +'#spenibus_archive > .list > .row:hover > * {'
         +'background-color:#FDA;'
      +'}'
   );


   // display box
   var box = SGL.displayBox('spenibus_archive');


   // store stuff
   var vars = {};


   // get media id
   var m = loc.pathname.match(/\/video\/(\d+)/i);


   // no media id found, abort
   if(!m) {
      return;
   }


   // store media id
   vars.mediaId = m[1];


   // get media
   GM_xmlhttpRequest({
      url    : 'https://www.hitbox.tv/api/media/video/'+vars.mediaId+'?showHidden=true',
      method : 'GET',
      onload : mediaHandler,
   });


   // media handler
   function mediaHandler(xhr) {

      vars.cfg = JSON.parse(xhr.responseText).video[0];
      var playlistConfig = JSON.parse(vars.cfg.media_profiles)[0];

      vars.playlistUrl = 'http://edge.hls.vods.hitbox.tv/static/videos/vods'
         +playlistConfig.url;

      // get playlist
      GM_xmlhttpRequest({
         url    : vars.playlistUrl,
         method : 'GET',
         onload : playlistParser,
      });
   }


   // playlist parser
   function playlistParser(xhr) {

      vars.filePrefix = vars.playlistUrl.match(/(.*\/)/)[1];

      vars.filesCount = 0;
      vars.duration   = 0;

      var lines = xhr.responseText.split('\n');
      vars.files = [];

      for(var i=0; i<lines.length; ++i) {

         var m = lines[i].match(/#EXTINF:([\d\.]+)/i);

         if(m) {

            ++vars.filesCount;

            // millisecond precision
            var d = Math.round(parseFloat(m[1]) * 1000);

            vars.duration += d;

            vars.files.push({
               duration : d,
               file     : lines[++i],
            });
         }
      }
      vars.filesCount = vars.files.length;


      // build list
      buildList();
   }


   // list builder
   function buildList() {


      var m = vars.cfg.media_date_added.match(/^(\d\d\d\d)-(\d\d)-(\d\d) (\d\d):(\d\d):(\d\d)/);
      var timestampStart = 0;
      if(m) {
         timestampStart = (new Date(
            parseInt(m[1]),
            parseInt(m[2])-1,
            parseInt(m[3]),
            parseInt(m[4]),
            parseInt(m[5]),
            parseInt(m[6])
         )).getTime();
      }

      var cleanTitle = vars.cfg.media_title.replace(/[\:*?"<>|/]/g, '_');


      var html = ''
         +'<div class="row header">'
            +'<div class="index">'+vars.filesCount+'</div>'
            +'<div class="start">'+timeFormat(timestampStart)+'</div>'
            +'<div class="duration">'+durationFormat(Math.round(vars.duration/1000))+'</div>'
            +'<div class="file"></div>'
         +'</div>';

      var index = 0;
      var fileStart = timestampStart;
      for(var i in vars.files) {

         var duration = durationFormat(Math.round(vars.files[i].duration/1000))

         var title = 'hitbox'+'-'
            +vars.cfg.media_user_name+'-'
            +timeFormat(fileStart)+'-'
            +cleanTitle;

         html += ''
            +'<div class="row header" title="'+title+'">'
               +'<div class="index">'+(++index)+'</div>'
               +'<div class="start">'+timeFormat(fileStart)+'</div>'
               +'<div class="duration">'+duration+'</div>'
               +'<div class="file">'
                  +'<a href="'+vars.filePrefix+vars.files[i].file+'">'
                     +'<span class="extra">'+title+'-</span>'
                     +vars.files[i].file
                  +'</a>'
               +'</div>'
            +'</div>';

         fileStart += vars.files[i].duration;
      }

      box.set('<div class="list">'+html+'</div>');
   }
}
window.addEventListener('DOMContentLoaded', archives, false);

jeuxvideocom.user.js
// ==UserScript==
// @name        jeuxvideocom
// @namespace   greasemonkey@spenibus
// @include     http*://jeuxvideo.com/*
// @include     http*://*.jeuxvideo.com/*
// @version     20150214-1728
// @require     spenibus-greasemonkey-lib.js
// @grant       GM_xmlhttpRequest
// ==/UserScript==




/***************************************************************** shorthands */
var SGL = spenibus_greasemonkey_lib;
var loc = document.location;




/******************************************************* video download links */
function videoDownloadLinks() {


   // check url
   if(loc.pathname.substr(0,8) != '/videos/') {

      // wrong url, abort
      return;
   }


   // display box
   var box = SGL.displayBox('spenibus_videoLinks');
   box.set('building links list');


   // vars storage
   var vars = {};


   // css
   vars.css = ''
      +'#spenibus_videoLinks * {'
         +'color:#000;'
         +'font-size:100%;'
      +'}'
      // list
      +'#spenibus_videoLinks > .list {'
         +'display:table;'
         +'white-space:nowrap;'
      +'}'
      // list rows
      +'#spenibus_videoLinks > .list > * {'
         +'display:table-row;'
      +'}'
      // list cells
      +'#spenibus_videoLinks > .list > * > * {'
         +'display:table-cell;'
         +'padding:0 4px;'
      +'}'
      // cell: label
      +'#spenibus_videoLinks > .list > * > *.label {'
         +'text-align:right;'
      +'}'
      // hide cells, except first cell of each row
      +'#spenibus_videoLinks > .list > * > * ~ * {'
         +'display:none;'
      +'}'
      // show cells on hover
      +':hover > #spenibus_videoLinks > .list > * > * ~ * {'
         +'display:table-cell;'
      +'}';




   //*********************************************************** get config data
   (function(){
      var node = document.querySelector('meta[property="og:video"]');

      vars.videoId = node.getAttribute('content').match(/(\d+)$/)[1];
      vars.cfgUrl  = '//'+loc.hostname+'/contenu/medias/video.php?q=config&id='+vars.videoId;

      box.set('looking up <a href="'+vars.cfgUrl+'">config</a>');

      GM_xmlhttpRequest({
         method : 'GET',
         url    : vars.cfgUrl,
         onload : function(r) {
            var data = JSON.parse(r.responseText);
            listBuild(data);
         }
      });
   })();




   //***************************************************************** make list
   var listBuild = function(data) {

      var list = '';
      for(var i in data.sources) {

         var item = data.sources[i];

         list += ''
            +'<div>'
               +'<div class="label">'+item.label+'</div>'
               +'<div class="file"><a href="'+item.file+'">'+item.file+'</a></div>'
            +'</div>';


         // css
         SGL.css(vars.css);


         // display
         box.set(''
            +'<a href="'+vars.cfgUrl+'">config</a>'
            +'<div class="list">'+list+'<div>'
         );
      }
   }
}
window.addEventListener('load', videoDownloadLinks, false);

stackoverflow.user.js
// ==UserScript==
// @name        stackoverflow
// @namespace   stackoverflow@spenibus
// @include     http*://stackoverflow.com/*
// @include     http*://*.stackoverflow.com/*
// @version     20150923-0255
// @require     spenibus-greasemonkey-lib.js
// @grant       unsafeWindow
// ==/UserScript==


/*******************************************************************************
first version: 20150807-0012
*******************************************************************************/




/***************************************************************** shorthands */
var SGL = spenibus_greasemonkey_lib;
var loc = document.location;



function fixedRound(val, len) {
    var p = Math.pow(10, len);
    return (Math.round(val * p) / p).toFixed(len);
}


/***************************************************************** flag stats */
if(loc.pathname.substr(0,19) =='/users/flag-summary') {

    window.addEventListener('DOMContentLoaded', function(){

        // pattern
        var r = /^\s*(\d+)(.*)/gm

        // collect counts
        var stats = {};

        // get flags text
        var n = document.getElementById('flag-stat-info-table');
        var s = n.textContent;

        // matches holder
        var m;


        // init to zero
        for(
            var i of [
                'moderator attention flags',
                'waiting for review',
                'deemed helpful',
                'declined',
                'disputed',
            ]
        ) {
            stats[i] = 0;
        }


        while(m = r.exec(s)) {

            if(!stats[m[2]]) {
                stats[m[2]] = 0;
            }

            stats[m[2]] += parseInt(m[1]);
        }

        var processedCount = (stats['deemed helpful'] + stats['disputed'] + stats['declined']) || 0;

        var factorProcessed = 100 / (processedCount || 1);
        var factorTotal     = 100 / (stats['moderator attention flags'] || 1); // (stats['deemed helpful'] + stats['declined']);

        var extraFlagsCount = stats['deemed helpful'] - stats['declined'];
        var extraFlags      = Math.floor(extraFlagsCount/10);

        var lost = stats['moderator attention flags']
            - stats['deemed helpful']
            - stats['disputed']
            - stats['declined']
            - stats['waiting for review'];

        var ns = document.createElement('div');
        ns.setAttribute('style', 'border:1px solid #000; padding:4px; margin:4px;');
        ns.innerHTML = ''
            +'<table id="spenibus_flags_stats">'
                +'<tr>'
                    +'<td colspan="99">Flags stats</td>'
                +'</tr>'
                +'<tr>'
                    +'<td>'+processedCount+'</td>'
                    +'<td>'+fixedRound(factorTotal * stats['moderator attention flags'], 2)+'%</td>'
                    +'<td>'+stats['moderator attention flags']+'</td>'
                    +'<td>total</td>'
                +'</tr>'
                +'<tr style="color:#080;">'
                    +'<td>'+fixedRound(factorProcessed * stats['deemed helpful'], 2)+'%</td>'
                    +'<td>'+fixedRound(factorTotal * stats['deemed helpful'], 2)+'%</td>'
                    +'<td>'+stats['deemed helpful']+'</td>'
                    +'<td>helpful</td>'
                +'</tr>'
                +'<tr style="color:#880;">'
                    +'<td>'+fixedRound(factorProcessed * stats['disputed'], 2)+'%</td>'
                    +'<td>'+fixedRound(factorTotal * stats['disputed'], 2)+'%</td>'
                    +'<td>'+stats['disputed']+'</td>'
                    +'<td>disputed</td>'
                +'</tr>'
                +'<tr style="color:#800;">'
                    +'<td>'+fixedRound(factorProcessed * stats['declined'], 2)+'%</td>'
                    +'<td>'+fixedRound(factorTotal * stats['declined'], 2)+'%</td>'
                    +'<td>'+stats['declined']+'</td>'
                    +'<td>declined</td>'
                +'</tr>'
                +'<tr style="color:#888;">'
                    +'<td></td>'
                    +'<td>'+fixedRound(factorTotal * stats['waiting for review'], 2)+'%</td>'
                    +'<td>'+stats['waiting for review']+'</td>'
                    +'<td>waiting</td>'
                +'</tr>'
                +'<tr style="color:#888;">'
                    +'<td></td>'
                    +'<td>'+fixedRound(factorTotal * lost, 2)+'%</td>'
                    +'<td>'+lost+'</td>'
                    +'<td>lost</td>'
                +'</tr>'
                +'<tr>'
                    +'<td colspan="99" class="sep">extra flags<br />'+extraFlags+' ('+extraFlagsCount+')</td>'
                +'</tr>'
                +'<tr>'
                    +'<td colspan="99" class="sep">helpful/declined ratio<br />'
                        +fixedRound(stats['deemed helpful']/(stats['declined']||1), 2)+'</td>'
                +'</tr>'
            +'</table>';

        n.parentNode.insertBefore(ns, n);

        SGL.css(''
            +'#spenibus_flags_stats {'
                +'width:100%;'
                +'font-weight:bold;'
                +'text-align:right;'
                +'margin:auto;'
            +'}'
            +'#spenibus_flags_stats td {'
                +'padding:0 4px;'
                +'text-align:right;'
            +'}'
            +'#spenibus_flags_stats td:nth-child(2n) {'
                +'background-color:rgba(0,0,0,0.05);'
            +'}'
            +'#spenibus_flags_stats .sep {'
                +'border-top:1px solid #888;'
            +'}'
        );

    }, false);
}




/************************************************************************ EOF */

youtube.user.js
// ==UserScript==
// @name        youtube
// @namespace   greasemonkey@spenibus
// @include     http*://youtube.com/*
// @include     http*://*.youtube.com/*
// @version     20141118-1248
// @require     spenibus-greasemonkey-lib.js
// @grant       unsafeWindow
// @grant       GM_xmlhttpRequest
// ==/UserScript==


/*******************************************************************************
creation: 2010-11-12 00:00 +0000
  update: 2014-11-18 12:48 +0000

TODO
   - handle "protected" videos, maybe
*******************************************************************************/




/***************************************************************** shorthands */
var SGL = spenibus_greasemonkey_lib;
var loc = document.location;




/*******************************************************************************
********************************************************************************
********************************************************************************
******************************* disable ajax page loading (history.pushState) */
if(unsafeWindow.ytspf) {
   unsafeWindow.ytspf.enabled = false;
}




/*******************************************************************************
********************************************************************************
********************************************************************************
***************************************************************** video links */
function videoLinks() {


   // not watch page, abort
   if(loc.pathname != '/watch') {
      return;
   }


   // init box and css
   SGL.css(''
      +'#spenibus_videoLinks {'
         +'display:table;'
         +'white-space:nowrap;'
         +'text-align:right;'
      +'}'
      +'#spenibus_videoLinks > div {'
         +'display:table-row;'
      +'}'
      +'#spenibus_videoLinks > div > div {'
         +'display:table-cell;'
         +'padding:0;'
      +'}'
      // all cells except first
      +'#spenibus_videoLinks > div > div ~ div {'
         +'padding-left:8px;'
      +'}'
      // cell: link
      +'#spenibus_videoLinks > div > div:last-child {'
         +'text-align:left;'
      +'}'
      // hide cells (except first)
      +'#spenibus_videoLinks > div > div ~ div {'
         +'display:none;'
      +'}'
      // hover: show cells
      +':hover > #spenibus_videoLinks > div > div ~ div {'
         +'display:table-cell;'
      +'}'
   );
   var box = SGL.displayBox('spenibus_videoLinks');


   //********************************************** missing configuration, abort
   if(!unsafeWindow.ytplayer) {
      box.set('ytplayer not found');
      return;
   }


   //********************************************************** prepare dash url
   var dashurl = (function() {
      if(unsafeWindow.ytplayer.config.args.dashmpd) {
         var tmp = unsafeWindow.ytplayer.config.args.dashmpd.match(/manifest\/dash\/(.*)/i)[1].split('/');
         var url = '';
         for(i=0; i<tmp.length; i=i+2) {
            url += '&'+unescape(tmp[i])+'='+unescape(tmp[i+1]);
         }
         url = url.substr(1);
         url += '&ratebypass=yes';
         return url;
      }
      return '';
   })();


   //****************************************************************** video id
   var videoId = (function(){
      box.set('fetching video id');

      // video id
      var tmp = document.querySelector('meta[itemprop=videoId]');
      tmp = tmp && tmp.getAttribute
         ? tmp.getAttribute('content')
         : false;

      return tmp;
   })();


   //****************************************************************** duration
   var duration = (function(){
      box.set('fetching duration');

      var tmp = document.querySelector('meta[itemprop=duration]');
      tmp = tmp.content
         ? tmp.content.match(/^PT(\d+)M(\d+)S$/i)
         : [0,0];

      var h = Math.floor(tmp[1]/60);
      var m = tmp[1]%60;
      var s = tmp[2];

      return ('0'+h).slice(-2)+':'+('0'+m).slice(-2)+':'+('0'+s).slice(-2);
   })();


   //************************************************************ published time
   var published = (function(){
      box.set('fetching published time');

      var tmp = {};

      if(!videoId) {
         return tmp;
      }

      //var tmpD = document.querySelector('span#eow-date');

      var xml = GM_xmlhttpRequest({
         method:      "GET",
         url:         "http://gdata.youtube.com/feeds/api/videos/"+videoId,
         synchronous: true,
      });

      var parser = new DOMParser();
      var doc = parser.parseFromString(xml.responseText, "application/xml");

      tmp.date = doc.querySelector('published');

      tmp.date = tmp.date && tmp.date.textContent
         ? new Date(tmp.date.textContent)
         : false;

      tmp.formatted = tmp.date
         ? tmp.date.getUTCFullYear()
            +''+('0'+(tmp.date.getUTCMonth()+1)).slice(-2)
            +''+('0'+tmp.date.getUTCDate()).slice(-2)
            +'-'+('0'+tmp.date.getUTCHours()).slice(-2)
            +''+('0'+tmp.date.getUTCMinutes()).slice(-2)
            +''+('0'+tmp.date.getUTCSeconds()).slice(-2)
            +'-UTC'
         : '';

      return tmp;
   })();


   //********************************************************************** user
   var user = (function(){
      box.set('fetching user');

      var tmp = document.querySelector('span[itemprop=author] > link[itemprop=url]');
      tmp = tmp.href
         ? tmp.href.match(/(?:user|channel)\/(.*)/)
         : ['',''];
      tmp =  tmp[1];

      return tmp;
   })();


   //********************************************************************* title
   var videoTitle = (function(){
      box.set('fetching title');

      var tmp = unsafeWindow.ytplayer.config.args.title
         ? unsafeWindow.ytplayer.config.args.title
         : '';

      return tmp;
   })();


   //*************************************************************** build title
   var title = (function(){
      box.set('building title');

      var str = user+' - '+published.formatted+' - '+videoTitle;

      // sanitize: special chars
      str = str.replace(/[\\\/\|\*\?<>:"]/g, '-');

      // sanitize: excessive spaces
      str = str.replace(/\s+/g, ' ');

      return str;
   })();


   //***************************************************** mimetype to extension
   box.set('building ext');
   var mimeToExt = {
      'video/webm'  : 'webm',
      'video/mp4'   : 'mp4',
      'video/x-flv' : 'flv',
      'video/3gpp'  : '3gp',
      'audio/mp4'   : 'm4a',
      'audio/webm'  : 'ogg',
   };


   //************************************************* fmt: number to resolution
   var numberResolution = (function(){
      box.set('building numRes');

      var list = unsafeWindow.ytplayer.config.args.fmt_list;
      if(!list) {
         return;
      }

      var out = {};
      var items = list.split(',');
      for(var i in items) {
         var tmp = items[i].split('/');
         out[tmp[0]] = tmp[1];
      }
      return out;
   })();


   //************************************************************* prepare items
   var items = (function(){
      box.set('preparing items');

      var obj = {
         data  : {},
         order : [],
      };

      var src = '';
      if(unsafeWindow.ytplayer.config.args.url_encoded_fmt_stream_map) {
         src += (src ? ',' : '')+unsafeWindow.ytplayer.config.args.url_encoded_fmt_stream_map;
      }
      if(unsafeWindow.ytplayer.config.args.adaptive_fmts) {
         src += (src ? ',' : '')+unsafeWindow.ytplayer.config.args.adaptive_fmts;
      }

      if(src) {
         var map = src.split(',');

         for(var i=0; i<map.length; i++) {
            var args = map[i].split('&');

            // get data
            var data = {};
            for(var n in args) {
               var tmp = args[n].split('=');
               data[tmp[0]] = unescape(tmp[1]);
            }

            // create item
            var item = {};
            item.itag    = data.itag;
            item.title   = title;
            item.ext     = mimeToExt[data.type.split(';')[0]];
            item.size    = data.size    ? data.size    : numberResolution[data.itag];
            item.bitrate = data.bitrate ? data.bitrate : 0;
            item.weight  = data.clen    ? data.clen    : 0;

            // item url
            item.url  = unescape(data.url);

            // add signature, obsolete ?
            //item.url += data.s   ? '&signature='+encodeURIComponent(data.s)   : '';
            //item.url += data.sig ? '&signature='+encodeURIComponent(data.sig) : '';

            // add to main object
            obj.data[item.itag] = item;
            obj.order.push(item.itag);
         }
      }
      return obj;
   })();


   //********************************************************** build links list
   var html_items = (function(obj){
      box.set('building links list');
      var str = '';
      for(var i in obj.order) {
         var item = obj.data[obj.order[i]];
         if(item) {
            str += ''
               +'<div>'
                  +'<div>'+item.itag+'</div>'
                  +'<div>'+(item.size ? item.size : '-')+'</div>'
                  +'<div>'+(item.bitrate ? Math.round(item.bitrate/1024)+' kibps' : '-')+'</div>'
                  +'<div>'+(item.weight ? Math.round(item.weight/1024/1024)+' mio' : '-')+'</div>'
                  +'<div><a href="'+item.url+'">'
                  +'youtube - '+item.title+' - fmt-'+item.itag+'.'+item.ext+'</a></div>'
               +'</div>';
         }
      }
      return str;
   })(items);


   //******************************************************************** output
   box.set(''
      +'<div>'
         +'<div>fmt</div>'
         +'<div>res</div>'
         +'<div>bitrate</div>'
         +'<div>size</div>'
         +'<div>'+duration+'</div>'
      +'</div>'
      +html_items
   );
}
window.addEventListener('DOMContentLoaded', videoLinks, false);




/*******************************************************************************
********************************************************************************
********************************************************************************
************************************************************** proxfree links */
function proxfreeLinks() {


   // not watch page, abort
   if(loc.pathname != '/watch') {
      return;
   }


   // init box and css
   SGL.css(''
      // show only first line by default
      +'#spenibus_proxfreeLinks > div ~ div {'
         +'display:none;'
      +'}'
      // show all lines on parent hover
      +':hover > #spenibus_proxfreeLinks > div ~ div {'
         +'display:block;'
      +'}'
   );


   var box = SGL.displayBox('spenibus_proxfreeLinks');
   box.set('<div>pf</div>');


   // get proxfree
   GM_xmlhttpRequest({
      method: 'POST',
      url: 'http://fr.proxfree.com/request.php?do=go',
      data: 'get='+encodeURIComponent(loc.href),
      headers: {
         'Content-Type' : 'application/x-www-form-urlencoded',
      },
      onload: function(r) {
         var lnk = r.responseText.match(/(proxfree\.com\/permalink\.php\?.*\);)/i);
         if(lnk) {
            box.append(
               '<div><a href="http://fr.'+lnk[0]+'">pf:fr</a></div>'+
               '<div><a href="http://tx.'+lnk[0]+'">pf:tx</a></div>'
            );
         }
      }
   });
}
window.addEventListener('DOMContentLoaded', proxfreeLinks, false);




/*******************************************************************************
********************************************************************************
********************************************************************************
********************************************************** faster flash wmode */
function wmode() {

   var selectorMap = {
      watch : 'embed#movie_player',
      embed : '#player > embed',
   }


   // page type
   var page = loc.pathname.match(/^\/(watch|embed)/);


   // get player
   var playerNode = page
      ? document.querySelector(selectorMap[page[1]])
      : false;


   // no player, abort
   if(!playerNode) {
      return;
   }


   // set wmode
   playerNode.setAttribute('wmode', 'direct');


   // show success (watch page only)
   if(page[1] == 'watch') {
      var box = SGL.displayBox();
      box.set('<span style="color:green;">wm</span>');
   }
}
window.addEventListener('DOMContentLoaded', wmode, false);